xbmc/input/touch/generic/GenericTouchInputHandler.cpp

```diff
@@ -361,44 +363,40 @@ void CGenericTouchInputHandler::triggerDetectors(TouchInput event, int32_t point
   {
     case TouchInputAbort:
     {
-      for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
-        delete (*detector);
       m_detectors.clear();
       break;
     }
     
     case TouchInputDown:
     {
-      for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
-        (*detector)->OnTouchDown(pointer, m_pointers[pointer]);
+      for (auto const& detector : m_detectors)
+        detector->OnTouchDown(pointer, m_pointers[pointer]);
       break;
     }
     
     case TouchInputUp:
     {
-      for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
-        (*detector)->OnTouchUp(pointer, m_pointers[pointer]);
+      for (auto const& detector : m_detectors)
+        detector->OnTouchUp(pointer, m_pointers[pointer]);
       break;
     }
     
     case TouchInputMove:
     {
-      for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
-        (*detector)->OnTouchMove(pointer, m_pointers[pointer]);
+      for (auto const& detector : m_detectors)
+        detector->OnTouchMove(pointer, m_pointers[pointer]);
       break;
     }
 
     default:
       return;
   }
 
-  std::set<IGenericTouchGestureDetector*> finishedDetectors;
-  for (std::set<IGenericTouchGestureDetector*>::const_iterator detector = m_detectors.begin(); detector != m_detectors.end(); ++detector)
+  for (auto it = m_detectors.begin(); it != m_detectors.end(); )
```

Comment:
`remove_if` does not work with `std::set` because it is ordered and `remove_if` needs to reorder elements. I see no strong reason why `m_detectors` would need to be a `std::set` though. Not like you can try to push two identical `unique_ptr`s anyway. Should we make it a `std::vector`?